pinot-server/src/main/java/org/apache/pinot/server/starter/helix/FreshnessBasedConsumptionStatusChecker.java

```diff
@@ -0,0 +1,154 @@
+
+  // constructor parameters
+  private final InstanceDataManager _instanceDataManager;
+  private final Set<String> _consumingSegments;
+  private final Long _minFreshnessMs;
+
+  // helper variable
+  private final Set<String> _caughtUpSegments = new HashSet<>();
+
+  public FreshnessBasedConsumptionStatusChecker(InstanceDataManager instanceDataManager, Set<String> consumingSegments,
+      Long minFreshnessMs) {
+    _instanceDataManager = instanceDataManager;
+    _consumingSegments = consumingSegments;
+    _minFreshnessMs = minFreshnessMs;
+  }
+
+  private boolean isOffsetCaughtUp(StreamPartitionMsgOffset currentOffset, StreamPartitionMsgOffset latestOffset) {
+    if (currentOffset != null && latestOffset != null) {
+      if (currentOffset.compareTo(latestOffset) == 0) {
+        return true;
+      }
+      long currentOffsetLong = ((LongMsgOffset) currentOffset).getOffset();
+      long latestOffsetLong = ((LongMsgOffset) latestOffset).getOffset();
+      // Kafka's "latest" offset is actually the next available offset. Therefore it will be 1 ahead of the
+      // current offset in the case we are caught up.
+      // We expect currentOffset == latestOffset if no messages have ever been published. Both will be 0.
+      // Otherwise, we never expect currentOffset > latestOffset, but we allow this to be caught up in case
+      // it ever happens so we're not stuck starting up.
+      return currentOffsetLong >= latestOffsetLong - 1;
+    }
+    return false;
+  }
+
+  protected Long now() {
+    return System.currentTimeMillis();
+  }
+
+  public int getNumConsumingSegmentsNotReachedMinFreshness() {
+    for (String segName : _consumingSegments) {
+      if (_caughtUpSegments.contains(segName)) {
+        continue;
+      }
+      TableDataManager tableDataManager = getTableDataManager(segName);
+      if (tableDataManager == null) {
+        LOGGER.info("TableDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        continue;
+      }
+      SegmentDataManager segmentDataManager = null;
+      try {
+        segmentDataManager = tableDataManager.acquireSegment(segName);
+        if (segmentDataManager == null) {
+          LOGGER.info("SegmentDataManager is not yet setup for segment {}. Will check consumption status later",
+              segName);
+          continue;
+        }
+        if (!(segmentDataManager instanceof LLRealtimeSegmentDataManager)) {
+          // There's a possibility that a consuming segment has converted to a committed segment. If that's the case,
+          // segment data manager will not be of type LLRealtime.
+          LOGGER.info("Segment {} is already committed and is considered caught up.", segName);
+          _caughtUpSegments.add(segName);
+          continue;
+        }
+        LLRealtimeSegmentDataManager rtSegmentDataManager = (LLRealtimeSegmentDataManager) segmentDataManager;
+        Long now = now();
```

Comment:
(minor) Change boxed `Long` to primitive, same for other places

Reply:
good catch, done